### DISCUSSION TITLE
IV consumption for AES-GCM-128 within IMS-AKA (IPSec) SIP registrations

### DISCUSSION THREAD

**Discussion by Praveen3146** [strongswan/strongswan#2496] [Q&A]
Hi, 
I am trying to add 'aes-gcm-128' into one of our products using openSSL 3.0.7. we currently use 'aes-cbc-128'. 
End points or the user devices use our product as a proxy during the SIP registrations. Proxy is a B2B agent within the context of IMS network. For testing, Ixia is being used to trigger IMS-AKA SIP registration via the proxy. And in the future Strongswan will be used too. Algorithm negotiation and SA setup is successful. Inbound REGISTER (ESP) packet decryption using 'aes-gcm-128' on the proxy fails. IMO, I might not be using IV correctly. 

I need help in understanding how exactly 'IV' is consumed on both sides. I have currently tried to use 8, 12, 16 bytes after AAD bytes from the inbound ESP packet as an IV. 

Questions:
a) Do you think, i might need to construct 12 byte IV using a SALT(4-bytes) plus the 8-bytes from the ESP packet after AAD? 
b) How would a receiver(proxy in this case) know what/how the sender(Ixia in this case) constructed the IV exactly?
c) Can someone please explain how IV is produced and consumed in detail for IMS-AKA using aes-gcm-128 with Strongswan?

**Comment by tobiasbrunner** [comment]
> a) Do you think, i might need to construct 12 byte IV using a SALT(4-bytes) plus the 8-bytes from the ESP packet after AAD?

Yes, while the transmitted IV is 8 bytes, an additional 4-byte salt value, that's part of the ESP SA's key material, is added to construct the full nonce for AES-GCM, refer to [RFC 4106](https://datatracker.ietf.org/doc/html/rfc4106) for details.

> b) How would a receiver(proxy in this case) know what/how the sender(Ixia in this case) constructed the IV exactly?

There is only one way to do this that's compliant with the standard.

**Reply by Praveen3146** [reply]
Thank you for the response! ESP SA's key material - what 4 octets is this exactly? is it the first 4 octets in CK for aes-128-gcm? and first 4 octets in IK for aes-128-gmac? RFC just mentions at a high level and nothing in specific.

**Reply by tobiasbrunner** [reply]
Did you read section 8.1?

**Reply by Praveen3146** [reply]
Coming back to this part of the discussion. 

AES-GCM-ESP with a 128 bit key
      The KEYMAT requested for each AES-GCM key is 20 octets.  The first
      16 octets are the 128-bit AES key, and the remaining four octets
      are used as the salt value in the nonce.

Since the sender of ESP packet already generated the KEYMAT, how would receiver know which 4 bytes in the 16 bytes key (aes-gcm-128 uses 16 bytes encryption Key denoted as CK) to use as SALT? OR can receiver too generate KEYMAT? If so, how ?

**Reply by tobiasbrunner** [reply]
Of course it has to generate the same KEYMAT. How else would it be able to decrypt the packets? How it does so depends on your keying protocol (you already know where to look for how IKEv2 does it). And the 4 bytes don't come out of the 16-byte AES key. As is described right there(!), the generated KEYMAT is 20 bytes, the first 16 bytes are the AES key, the remaining 4 the salt that's used to build the nonce.

**Reply by Praveen3146** [reply]
agreed, so the SALT generation for IMS-AKA is probably based on the standards I shared in previous responses (unlike IKEv2). reattaching again below. So, I might need to use KDF to come up with 4 bytes SALT as described below using the secret keys and other inputs. Use the produced SALT along with 8 bytes IV in the packet as AES-GCM nonce of 12 bytes. 

https://www.etsi.org/deliver/etsi_ts/133200_133299/133203/17.01.00_60/ts_133203v170100p.pdf

Integrity Keys: If the selected authentication algorithm is HMAC-SHA-1-96 then IKESP is obtained from IKIM by appending 32 zero bits to the end of IKIM to create a 160-bit string. If selected authentication algorithm is AES-GMAC as specified in RFC 4543 [74] with 128 bit key then IKESP = IKIM. The salt value specified in Section 3.2 of RFC 4543 [74] shall be derived using the key derivation function KDF defined in Annex B of TS 33.220 [66]. The input Key to the KDF function shall be equal to the concatenation of CKIM and IKIM: CKIM || IKIM. The input S to the KDF function shall be formed from the following parameters: - FC = 0x58. - P0 = "AES_GMAC_SALT" . - L0 = length of the string “AES_GMAC_SALT” (i.e. 0x00 0x0D). The salt value shall consist of the 32 least significant bits of the 256 bits of the KDF output. "Hmac-sha-1-96" is not recommended.

Encryption Keys: If selected encryption algorithm is AES-CBC as specified in RFC 3602 [22] with 128 bit key then CKESP = CKIM . If selected encryption algorithm is AES-GCM as specified in RFC 4106 [73] with 128 bit key then CKESP = CKIM. The salt value specified in Section 4 of RFC 4106 [73] shall be derived using the key derivation function KDF defined in Annex B of TS 33.220 [66]. The input Key to the KDF function shall be equal to the concatenation of CKIM and IKIM: CKIM || IKIM. The input S to the KDF function shall be formed from the following parameters: - FC = 0x59 - P0 = “AES_GCM_SALT” - L0 = length of the string “AES_GCM_SALT” (i.e. 0x00 0x0C) The salt value shall consist of the 32 least significant bits of the 256 bits of the KDF output. "aes-cbc" is not recommended.

**Comment by Praveen3146** [comment]
Yes, read it after I posted the comment. Question is what would be the
inputs to the PRF on the receiver end.

1.

IKE makes use of a pseudo-random function (PRF) to derive keying
   material.  The PRF is used iteratively to derive keying material of
   arbitrary size, called KEYMAT.  Keying material is extracted from the
   output string without regard to boundaries.

   The size of the KEYMAT for the AES-GCM-ESP MUST be four octets longer
   than is needed for the associated AES key.  The keying material is
   used as follows:

   AES-GCM-ESP with a 128 bit key
      The KEYMAT requested for each AES-GCM key is 20 octets.  The first
      16 octets are the 128-bit AES key, and the remaining four octets
      are used as the salt value in the nonce.


2.
Example I found:
What would be the 'key' and 'input' so receiver can derive the same SALT as
the Sender?

void prf(unsigned char *key, unsigned char *input, int input_len,
unsigned char *output) {
    unsigned char buffer[SHA256_DIGEST_LENGTH];

    // Concatenate key and input
    unsigned char combined[SHA256_DIGEST_LENGTH + input_len];
    memcpy(combined, key, SHA256_DIGEST_LENGTH);
    memcpy(combined + SHA256_DIGEST_LENGTH, input, input_len);

    // Compute SHA-256 hash
    SHA256(combined, sizeof(combined), buffer);

    // Copy first 16 bytes of the hash as output
    memcpy(output, buffer, 16);
}

int main() {
    unsigned char key[SHA256_DIGEST_LENGTH] = "mysecretkey"; //
Replace with a secure key
    unsigned char input[] = "somedata";
    unsigned char output[16];

    prf(key, input, strlen((char *)input), output);

    // Print the output in hex format
    for (int i = 0; i < 16; i++) {
        printf("%02x", output[i]);
    }
    printf("\n");

    return 0;
}

**Reply by tobiasbrunner** [reply]
> Question is what would be the inputs to the PRF on the receiver end.

That has nothing to do with the selected encryption algorithm. Read [RFC 7296, section 2.17](https://datatracker.ietf.org/doc/html/rfc7296#section-2.17).

**Reply by Praveen3146** [reply]
1. RFC 7296 section 2.17 is relevant for IKE. IPSec in IMS-AKA  doesn't involve key exchanges like described in this section. 
There is no IKE_AUTH exchange, instead the SAs are created based on SIP negotiation.

2. I've also come across this for IMS-AKA key expansion. Just not sure which one to follow yet.. 
    That is why I need help in understanding what/how Strongswan handles IV consumption on both ends for IMS-AKA IPSec specifically. 
 
https://www.etsi.org/deliver/etsi_ts/133200_133299/133203/17.01.00_60/ts_133203v170100p.pdf
 
Integrity Keys: If the selected authentication algorithm is HMAC-SHA-1-96 then IKESP is obtained from IKIM by appending 32 zero bits to the end of IKIM to create a 160-bit string. If selected authentication algorithm is AES-GMAC as specified in RFC 4543 [74] with 128 bit key then IKESP = IKIM. The salt value specified in Section 3.2 of RFC 4543 [74] shall be derived using the key derivation function KDF defined in Annex B of TS 33.220 [66]. The input Key to the KDF function shall be equal to the concatenation of CKIM and IKIM: CKIM || IKIM. The input S to the KDF function shall be formed from the following parameters: - FC = 0x58. - P0 = "AES_GMAC_SALT" . - L0 = length of the string “AES_GMAC_SALT” (i.e. 0x00 0x0D). The salt value shall consist of the 32 least significant bits of the 256 bits of the KDF output. "Hmac-sha-1-96" is not recommended.
 
Encryption Keys: If selected encryption algorithm is AES-CBC as specified in RFC 3602 [22] with 128 bit key then CKESP = CKIM . If selected encryption algorithm is AES-GCM as specified in RFC 4106 [73] with 128 bit key then CKESP = CKIM. The salt value specified in Section 4 of RFC 4106 [73] shall be derived using the key derivation function KDF defined in Annex B of TS 33.220 [66]. The input Key to the KDF function shall be equal to the concatenation of CKIM and IKIM: CKIM || IKIM. The input S to the KDF function shall be formed from the following parameters: - FC = 0x59 - P0 = “AES_GCM_SALT” - L0 = length of the string “AES_GCM_SALT” (i.e. 0x00 0x0C) The salt value shall consist of the 32 least significant bits of the 256 bits of the KDF output. "aes-cbc" is not recommended.

**Reply by tobiasbrunner** [reply]
> RFC 7296 section 2.17 is relevant for IKE. IPSec in IMS-AKA doesn't involve key exchanges like described in this section.
> There is no IKE_AUTH exchange, instead the SAs are created based on SIP negotiation.

Then read the relevant standards that define how keys for IPsec are derived in your scenario. You might have already found it.


>  That is why I need help in understanding what/how Strongswan handles IV consumption on both ends for IMS-AKA IPSec specifically.

You seem confused as you refer to strongSwan, which is an IKE daemon and as you pointed out completely irrelevant to this. And no idea what you mean with "IV consumption".

**Reply by Praveen3146** [reply]
Correct, I am reading all the related standards and did come across multiple of them talking about how IV is constructed and used for aes-gcm, there is definitely high level of confusion. To help clear the confusion, Since Strongswan may support aes-gcm for IKE and IMS-AKA, there might be folks using it for IMS-AKA IPSec SIP registrations etc. And my expectation from Strongswan developer community is to get clarity on how IV is consumed (constructed and used on both sides) within IPSec IMS-AKA for aes-gcm specifically. 
would be nice to share with me the specifics if possible.

**Reply by tobiasbrunner** [reply]
Now I'm confused. I already shared with you the RFCs that describe how ESP uses AES-GCM (i.e. how the nonce is constructed from the IV in the packet and the salt from the key material) and how the key material is created (during IKEv2). What more can I do? And not sure why you'd think strongSwan has anything to do with IMS-AKA.
